Versioning and Stability for OpenTelemetry (#1291)
open-telemetry/opentelemetry-specification: specification/versioning-and-stability.md

# Versioning and stability for OpenTelemetry clients
<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Design goals](#design-goals)
- [Signal lifecycle](#signal-lifecycle)
  * [Lifecycle stages](#lifecycle-stages)
  * [Stability](#stability)
  * [Deprecation](#deprecation)
  * [Removal](#removal)
  * [A note on replacing signals](#a-note-on-replacing-signals)
- [Version numbers](#version-numbers)
  * [Major version bump](#major-version-bump)
  * [Minor version bump](#minor-version-bump)
  * [Patch version bump](#patch-version-bump)
- [Long Term Support](#long-term-support)
  * [API support](#api-support)
  * [SDK Support](#sdk-support)
  * [Contrib Support](#contrib-support)
- [OpenTelemetry GA](#opentelemetry-ga)

<!-- tocstop -->

This document defines the stability guarantees offered by the OpenTelemetry clients, along with the rules and procedures for meeting those guarantees.

In this document, the term OpenTelemetry specifically refers to the OpenTelemetry clients. It does not refer to the specification or the Collector.

> Language implementations MUST take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met. This proposal shall be placed in the root of each repo and named `VERSIONING.MD`.

Reviewer: Do we still need this paragraph? If yes, maybe it will help to think of this as a mandatory section, more than a proposal.

Author: Thanks, I updated the language here. We still need this paragraph IMHO, to make the requirement that VERSIONING.MD documents must exist in each repo.

**Deprecated –** this signal has been replaced, but MUST maintain the same stability guarantees.

**Removed -** a deprecated signal is no longer supported, and MUST be removed from the release.

All signal components MAY become stable together, or MAY transition to stability component-by-component in the following order: API, SDK, Semantic Conventions, Contrib.

When transitioning from experimental to stable to deprecated, signals MUST NOT move or otherwise break how they are imported by users. Terms which denote stability, such as "experimental," MUST NOT be used as part of a directory or import name.

Package **version numbers** MAY include a suffix, such as -alpha, -beta, -rc, or -experimental, to differentiate stable and experimental packages.

### Stability

Once a signal component is marked as stable, the following rules MUST apply until the end of that signal’s existence.

**API Stability -**
Backward-incompatible changes to API packages MUST NOT be made unless the major version number is incremented. All existing API calls MUST continue to compile and function against all future minor versions of the same major version.

> ABI compatibility for the API is RECOMMENDED, and MAY be offered on a language by language basis.

Reviewer 2: Is a recommendation the same as a simpler `Languages SHOULD offer ABI compatibility for the API.`?

Author: Yes, RECOMMENDED, MAY, and SHOULD mean the same thing from the perspective of the [IETF](https://www.ietf.org/rfc/rfc2119.txt).

I like your wording better. :)

> ABI compatibility for the API is RECOMMENDED, and MAY be offered on a language by language basis.

Reviewer 3: Would it be better to caveat this that "Languages which ship binary artifacts SHOULD offer ABI compatibility" or even MUST?

If we take the notion that "API compatible" = source compiles and "ABI compatible" = binaries link, then I think the nuance here is more around the distribution mechanism.   No binary distribution => ABI doesn't exist.

Author: Thank you, much clearer. 

I've avoided saying MUST due to push back from maintainers. It is a bit difficult to make a sweeping, cross-language statement here.

**SDK Stability -**
Public portions of SDK packages MUST remain backwards compatible. There are two categories of public features: **plugin interfaces** and **constructors**. Examples of plugin include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders.

ABI compatibility for public portions of the SDK are RECOMMENDED, and MAY be offered on a language by language basis.

**Semantic Conventions Stability -**

> Semantic Conventions MUST NOT be removed once they are stable. New conventions MAY be added to replace usage of older conventions, but the older conventions MUST NOT be removed. Older conventions MUST be marked as deprecated when they are replaces by newer conventions.

Reviewer 2: ```suggestion
Semantic Conventions MUST NOT be removed once they are stable. New conventions MAY be added to replace usage of older conventions, but the older conventions MUST NOT be removed. Older conventions MUST be marked as deprecated when they are replaced by newer conventions.
```

> Semantic Conventions MUST NOT be removed once they are stable. New conventions MAY be added to replace usage of older conventions, but the older conventions MUST NOT be removed. Older conventions MUST be marked as deprecated when they are replaces by newer conventions.

Reviewer 3: nit: spelling:  s/replaces/replaced/

**Contrib Stability -**
Plugins and instrumentation MUST be kept up to date, and MUST released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Public portions of contrib packages (constructors, configuration, interfaces) MUST remain backwards compatible.

> ABI compatibility for public portions of contrib packages is RECOMMENDED, and MAY be offered on a language by language basis.

Reviewer 3: Nit: Could we expand this to encompass the difficulty of a plugin's ecosystem vs. the language itself?

E.g. a prometheus plugin can only offer the guaranteees allowed by the prometheus downstream dependencies it relies on.

Author: That is a good point. I added an exception, let me know if it works.

Contrib stability is going to be a rough ride...

> Telemetry produced by contrib instrumentation MUST remain stable and backwards compatible, to avoid breaking alerts and dashboard. Existing telemetry MUST NOT be mutated or removed without a major version bump. Additional telemetry MAY be added. This includes additional spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.

Reviewer 3: Nit:  I think it's worth leaving a placeholder after this (which should be discussed outside this PR) about what to do when a contrib plugin has an instable release and we want to support BOTH versions of that ecosystem.   Hopefully this never happens, but....  It'd be nice to know what to do if, e.g. Prometheus decided to have some kind of major breaking change that made us unable to support our previous version in a stable/compatible way with a new version.

Note: I'm picking on prometheus just because it's the biggest non-OTel plugin I can think of, not because I expect this to happen.

Author: Yeah, generally speaking we want to maintain support for v3 of Foo, even when v4 of Foo becomes available. Added a clarification.

In general, immutability is our friend here. The less we mutate the old stuff, the better.

Author: Added additional language to the exception to clarify this.

### Deprecation

Signals MAY eventually be replaced. When this happens, they are marked as deprecated.

> Signals SHALL only marked as deprecated when the replacement becomes stable. Deprecated code MUST abide by the same support guarantees as stable code.

Reviewer 2: ```suggestion
Signals SHALL only be marked as deprecated when the replacement becomes stable. Deprecated code MUST abide by the same support guarantees as stable code.
```

* New backward-compatible functionality added to any component.
* Breaking changes to internal SDK components.
* Breaking changes to experimental signals.
* New experimental signals are added.
* Experimental signals become stable.
* Stable signals are deprecated.

### Patch version bump

Patch versions make no changes which would require recompilation or potentially break application code. The following are examples of patch fixes.

* Bug fixes which don't require minor version bump per rules above.
* Security fixes.
* Documentation.

Currently, OpenTelemetry does NOT have plans to backport bug and security fixes to prior minor versions. Security and bug fixes MAY only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.

## Long Term Support

![long term support](../internal/img/long-term-support.png)

### API support

Major versions of the API MUST be supported for a minimum of **three years** after the release of the next major API version. API support is defined as follows.

API stability, as defined above, MUST be maintained.

> A version of the SDK which supports the latest minor version of the last major version of the API will continue to be maintained during LTS. Bug and security fixes MUST be backported. Additional feature development is NOT RECOMMENDED.

Reviewer: > Additional feature development is NOT RECOMMENDED.

This would cover adding new members to existing interfaces, correct?

Author: Yes; don't make the old dogs learn new tricks. :)

Contrib packages available when the API is versioned MUST continue to be maintained for the duration of LTS. Bug and security fixes will be backported. Additional feature development is NOT RECOMMENDED.

### SDK Support

> SDK stability, as defined above, will be maintained for a minimum of **one year** after after the release of the next major SDK version.

Reviewer 2: ```suggestion
SDK stability, as defined above, will be maintained for a minimum of **one year** after the release of the next major SDK version.
```

Reviewer 2: I'm not sure, but is there a situation where old versions of the API only work with old versions of the SDK?

Author: Yes, there is one: when the API makes a major version bump. In that case, a version of the SDK (and contrib) which works with the prior major version must be maintained for the duration of the API LTS.

### Contrib Support

> Contrib stability, as defined above, will be maintained for a minimum of **one year** after after the release of the next major version of a contrib package.

Reviewer 3: What is the plan to support drive-by-contributions to contrib?    E.g. let's say someone comes with a really meaty and great plugin for system X and then their life-choices change and they're no longer able to maintain this plugin in the community.   Now, contrib has a large and unmaintained portion of code that is a liability.   Do we expect other contrib authors/maintainers to jump in and learn that ecosystem to fix bugs?   Is contrib meant to be more "communtiy self-service"?

Author: Actually, contrib is the opposite in that sense. Packages are only added to contrib with the understanding that the OpenTelemetry SIG is now on the hook for maintaining them. This is both a stability and a security issue.

Packages maintained by third parties can be added to the [registry](http://opentelemetry.io/registry) for discoverability.

Reviewer 4: > OpenTelemetry SIG is now on the hook for maintaining them

I think this needs to be clarified in some official place. It was raised in maintainers meeting, and it was told that an issue will be opened to track "how to accept contrib and maintain.". It may be better to add TODO in this PR to that issue.

# Versioning and stability for OpenTelemetry clients
<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Design goals](#design-goals)
- [Signal lifecycle](#signal-lifecycle)
  * [Lifecycle stages](#lifecycle-stages)
  * [Stability](#stability)
  * [Deprecation](#deprecation)
  * [Removal](#removal)
  * [A note on replacing signals](#a-note-on-replacing-signals)
- [Version numbers](#version-numbers)
  * [Major version bump](#major-version-bump)
  * [Minor version bump](#minor-version-bump)
  * [Patch version bump](#patch-version-bump)
- [Long Term Support](#long-term-support)
  * [API support](#api-support)
  * [SDK Support](#sdk-support)
  * [Contrib Support](#contrib-support)
- [OpenTelemetry GA](#opentelemetry-ga)

<!-- tocstop -->

This document defines the stability guarantees offered by the OpenTelemetry clients, along with the rules and procedures for meeting those guarantees.

In this document, the terms "OpenTelemetry" and "language implementations" both specifically refer to the OpenTelemetry clients. These terms do not refer to the specification or the Collector in this document.

> Each language implementation MUST take these versioning and stability requirements, and produce a language-specific document which details how these requirements will be met. This document SHALL be placed in the root of each repo and named `VERSIONING.MD`.

Author: Sure, no problem.

**Removed -** a deprecated signal is no longer supported, and MUST be removed from the release.

All signal components MAY become stable together, or MAY transition to stability component-by-component in the following order: API, Semantic Conventions, API Contrib, SDK, SDK Contrib.

When transitioning from experimental to stable to deprecated, signals MUST NOT move or otherwise break how they are imported by users. Terms which denote stability, such as "experimental," MUST NOT be used as part of a directory or import name.

Package **version numbers** MAY include a suffix, such as -alpha, -beta, -rc, or -experimental, to differentiate stable and experimental packages.

### Stability

Once a signal component is marked as stable, the following rules MUST apply until the end of that signal’s existence.

**API Stability -**
Backward-incompatible changes to API packages MUST NOT be made unless the major version number is incremented. All existing API calls MUST continue to compile and function against all future minor versions of the same major version.

Languages which ship binary artifacts SHOULD offer ABI compatibility for the API.

**SDK Stability -**

> Public portions of SDK packages MUST remain backwards compatible. There are two categories of public features: **plugin interfaces** and **constructors**. Examples of plugins include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders.

Reviewer 2: @tedsuo Any thoughts on a very small part of the API being annotated as alpha and thus not stable yet? While navigating the world with separate artifacts / stability for metrics, I found myself possibly wanting that. https://github.com/open-telemetry/opentelemetry-java/pull/2330/files Maybe not, but wondering what the feasibility is vs this spec.

Author: Hmmm, not sure what you mean by a "small part"? Do you mean keeping the metrics API alpha while the Tracing API becomes stable? Or something else?

Reviewer 2: Yeah - so for example the Tracing SDK may reference the metrics API if it wants to provide monitoring of span processors. To be able to configure this, the Tracing SDK would need to expose a small API to allow configuring a `MeterProvider`. Cross cutting concerns :)

Author: Well, if a stable tracing SDK wants to add the metrics API to its public surface area, then the tracing SDK gets a major version bump when the metrics API breaks. So the options appear to be: 

* Keep the tracing SDK marked as experimental, only mark the trace API and trace API contrib as stable.
* Manage the relationship via a stable trace plugin interface (i.e. make the metrics bit a `SpanProcessor` plugin, rather than add a `MetricsProvider` to the surface of the tracing SDK).
* Refrain from integrating metrics into the tracing SDK until the metrics API is stable.

A related tricky question is instrumentation. Seems like metrics instrumentation should be kept in separate packages from tracing instrumentation until both are stable. I'm skeptical of "standardized metrics instrumentation" so that whole concept is a can of worms in general; nothing stable going on in that department any time soon.

Reviewer 2: I think you're saying that you don't like the idea of declaring parts of APIs as unstable, via an annotation in Java as an example. That sounds fine to me :)

Author: Yeah, I'd like us to at least start with some firm bright lines. If we can't do it then there's a smell of some kind. 

Longer term if there is a way to be more nuanced without giving up ABI compatibility or other important features I'm all ears. :)

Languages which ship binary artifacts SHOULD offer ABI compatibility for public portions of the SDK.

**Semantic Conventions Stability -**

> Semantic Conventions MUST NOT be removed once they are stable. New conventions MAY be added to replace usage of older conventions, but the older conventions MUST NOT be removed. Older conventions MUST be marked as deprecated when they are replaced by newer conventions.

Reviewer 5: What about non-required semantic attributes? Is it allowed for auto-instrumentation to provide them in one version and stop providing them in the next minor version? They are optional, after all, right?

Author: I think that fall under a different section, "contrib stability," in reference to the data produced by instrumentation. This section covers the stability of the conventions themselves - we can't remove the constants or anything else once they have been added to the conventions package. 

But to actually answer your question - we can't remove any data produced by instrumentation without a major version bump of that instrumentation library, as it could break dashboards and alerting. 

So, yes, they can be removed. But data should be deprecated first, with the actual deletions bundled up for release along with other major instrumentation changes, rather than have the breaking changes dribble out one by one.

Reviewer 5: But if attribute is optional, then all backends have to work with that attribute missing. Essentially you are saying "the moment auto-instrumentation adds some optional attribute for a particular instrumented library, it has to support is forever". What if next version of that library will not provide that data?

Also, what is the deprecation mechanism for semantic attributes? How can we communicate to our clients (meaning every user of auto-instrumentation) that these attribute from from instrumented library is deprecated?

**Contrib Stability -**

> Plugins and instrumentation MUST be kept up to date, and MUST released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Author: For the terminology, a couple points. I have this at the top:
> In this document, the terms "OpenTelemetry" and "language implementations" both specifically refer to the OpenTelemetry clients. These terms do not refer to the specification or the Collector in this document.

But I am thinking that needs to change. In general, we are sloppy about the term OpenTelemetry.

1) Use the term "OpenTelemetry clients" everywhere, rather than a short hand.
2) OpenTelemetry clients are defined as the API, SDK, Semantic Conventions, and Contrib packages, collectively. This could be made clearer in the overview.

Note that Contrib is part of "Official" OpenTelemetry; that is defined already under the definition of Contrib in the Overview. 3rd party plugins hosted elsewhere are not part of contrib. 

As far as MUST vs SHOULD, I do think it is reasonable that a SIG update all relevant Contrib packages as part of releasing a new version of the SDK, API, or Semantic Conventions. This should be done in a coordinated fashion – it is unfair to users to claim that a new version of the SDK is available when supported plugins do not work with it. We are not in the business of making breaking changes left and right; keeping Contrib packages up to date should not be a burden. 

If it is a burden, then we are breaking too many things. If we cannot keep up with our own changes, then 3rd party packages will definitely not be able to keep up, and users will be stuck on older versions of the SDK due to a lack of compatible plugins. It is a priority to avoid this situation. This is defined under Design Goals, let me know if that section is clear enough.

Likewise, instrumentation should always include the latest conventions and remove any API calls to deprecated methods. If this is not tracked in a coordinated fashion, it will become a mess. So, no waiting here either. Again, the API and conventions must remain stable and backwards compatible, so releasing a new version does not mean a massive rewrite of instrumentation is in order.

I feel that MUST makes this clearer than SHOULD. Let me know your thoughts.

> Plugins and instrumentation MUST be kept up to date, and MUST released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Reviewer 6: Since defining "OpenTelemetry" as clients in this doc is confusing, I think it's better to just call it OpenTelemetry clients or any other appropriate term. :+1:

> As far as MUST vs SHOULD, I do think it is reasonable that a SIG update all relevant Contrib packages as part of releasing a new version of the SDK, API, or Semantic Conventions.

I also expect it (official contrib packages are supporting new changes when released..) - but as it is "MUST", a delay of contrib implementation for any reason (e.g. lack of contributions or complexity of implementation in specific language) MUST block any release of spec changes - or we may break the promise("shortly after") - which is not desirable.

Actually, the concept of "contrib package" is being introduced in this doc - what are the example of contrib packages? Why should OpenTelemetry team need to maintain it, instead of enforcing the rules?

From previous section:

> Plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel, are referred to as Contrib packages.

From https://github.com/open-telemetry/opentelemetry-specification/pull/1291/files#r544742481

> [Contrib]  Packages are only added to contrib with the understanding that the OpenTelemetry SIG is now on the hook for maintaining them. This is both a stability and a security issue.

> Plugins and instrumentation MUST be kept up to date, and MUST released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Author: Thanks for the feedback, I rewrote the definition of Contrib packages and added examples for clarification.

I do believe that if we offer integrations, we must keep them up to date or we have broken a promise to our end users. If we say SHOULD, then we are weakening that promise. I agree that there could be extenuating circumstances; you make a good case for SHOULD. But... I have seen a pattern where maintainers may consider this code to be less important than the SDK code. It isn't. OpenTelemetry is only as good as it's instrumentation. We need to commit to maintaining this ecosystem. The fact that it may be difficult is not a sufficient reason to loosen the requirement. Backwards compatibility is also difficult to maintain; that does not make it optional. So SHOULD makes me concerned. I will discuss this point at the maintainer's meeting tomorrow and get more feedback. In the meantime, I added some clarifying language to this section. Thank you for the well reasoned discussion.

> Plugins and instrumentation MUST be kept up to date, and MUST released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Author: Based on the maintainers meeting today, we are going with SHOULD for now. But in general, it is clear that we need a more well defined plan for what we allow in contrib, how we maintain it, and how we treat 3rd-party plugins and instrumentation which we do not maintain.

The situation I want to avoid is one where we agree to take something into contrib, but that results in issues and PRs not being managed properly.

On the one hand, if these plugins are left to 3rd parties, then there is no "blessed" plugin and it is easier for alternatives to be generated by new maintainers. On the other hand, we are concerned about the security and reliability implications of packaging up and distributing 3rd-party plugins in the auto-installers and releases we provide. 

We'd prefer to have official plugins which are reliably maintained, but we do not yet have a clear plan for how to do this in a manner which addresses all concerns. We need a plan, but that is outside of the scope of this current proposal. So, leaving it as SHOULD for now.

If you have a good idea on how to address this, please consider making an OTEP. :)

### Removal

Support is ended by the removal of a signal from the release. The release MUST make a major version bump when this happens.

### A note on replacing signals

Note that we currently have no plans for creating a major version of OpenTelemetry past v1.0.

For clarity, it is still possible to create new, backwards incompatible versions of existing signals without actually moving to v2.0 and breaking support.

For example, imagine we develop a new, better tracing API - let's call it AwesomeTrace. We will never mutate the current tracing API into AwesomeTrace. Instead, AwesomeTrace would be added as an entirely new signal which coexists and interoperates with the current tracing signal. This would make adding AwesomeTrace a minor version bump, *not* v2.0. v2.0 would mark the end of support for current tracing, not the addition of AwesomeTrace. And we don't want to ever end that support, if we can help it.

This is not actually a theoretical example. OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing. We invented a new tracing API, but continue to support the old one.

## Version numbers

> OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinctions.

Reviewer 6: It would be better to group what's the distinction from semver. The current following section is to guide how to version packages altogether - not necessarily "distinctions" from semver.

Author: Maybe "clarifications" is a better term. We are not deviating from semver, but we have some rules as to how we follow it.

OpenTelemetry clients have four components: API, SDK, Semantic Conventions, and Contrib.

For the purposes of versioning, all code within a component MUST treated as if it were part of a single package, and versioned with the same version number, except for Contrib, which may be a collection of packages versioned separately.

* API packages for all signals MUST version together, across all signals. Signals MUST NOT have separate version numbers. There is one version number that applies to all signals that are included in the API release that is labeled with that particular version number.

> * SDK packages for all signals MUST version together, across all signals. Signals MUST NOT have separate version numbers. There is one version number that applies to all signals that are included in the SDK release that is labeled with that particular version number.

Reviewer 6: I'm not sure this is practical to use one version number for all signals.

My understanding is, one of the design goal is to have independent signals so that each signal can be packaged independently. Then.. any breaking changes in one signal require bumping major version of all packages, even others do not need any changes. This might be okay, but then why don't we have just one package for all signal? :confused: (note: this is kind of Rails model - ActiveRecord, ActiveModel... all are bumped altogether with Rails)

Or will I be able to mix different versions when using them? (e.g. `1.0.0` for a signal and `1.0.1` for another signal, as far as they can be used with same API package) If so, then what's the point of aligning the signal SDK versions?

Also note that signals may be in separate stages per spec (stable or experimental, for instance) - and it may be better to give 0.x or add pre-release tag to experimental signals. Enforcing one version (e.g. `1.2.3`) to all signal packages would make less use of semver actually.

BTW, as a user (and a developer as well), I prefer to have one SDK (reference) package for all signals for easier upgrade and maintenance actually, with annotating each module/function/class whatsoever with the stage.

Author: We definitely want as few version numbers flying around as possible, and no mixing and matching. There are a number of reasons for this. Stable signals may depend upon each other. Most implementations will offer some kind of unified API or SDK package, which pulls in all of the stable signals and provides additional connivence methods. Contrib instrumentation will contain multiple signals, and the SDK may wish to instrument itself. We provided tested and verified releases. All of this is supported by treating everything within a component as part of a single release with a single version number. The four components within a release (API, SDK, SemConv, Contrib) only have separate versions due to the different time scales and different audiences who depend upon them. 

Signals are in separate packages because otherwise they could not be developed experimentally without destabilizing existing stable packages.  Experimental signals cannot be brought into any unified API/SDK/Contrib package until they become stable. So, we end up with an architecture where each signal is in its own package, but as a user, it is still possible to bring in a single package which references everything stable. If you want to use a package while it is experimental, you have to reference it directly so that you are aware of what you are doing. Since these direct references do not break when the package becomes stable, we don't punish our beta users.

I agree, this is similar to how Rails handles versioning. Everything is well factored into separate components which are then wrapped up together wherever that creates convenience, and everything has the same version number so that you don't lose your mind. 

(BTW, the infographics attempt to describe these various relationships. I will make sure this information is also in the text, so that it is not missed by screen readers.)

Note that experimental signals may keep a 0.X version number. That is already defined in this section, under "exceptions."

* Semantic Conventions are a single package with a single version number.
* Each contrib package MAY have it's own version number.
* The API, SDK, Semantic Conventions, and contrib components are NOT REQUIRED to share a version number. For example, the latest version of `opentelemetry-python-api` may be at v1.2.3, while the latest version of `opentelemetry-python-sdk` may be at v2.3.1.
* Different language implementations are NOT REQUIRED to have matching version numbers. For example, it is fine to have `opentelemetry-python-api` at v1.2.8 when `opentelemetry-java-api` is at v1.3.2.
* Language implementations are NOT REQUIRED to match the version of the specification they implement. For example, it is fine for v1.8.2 of `opentelemetry-python-api` to implement v1.1.1 of the specification.

**Exception:** in some languages, package managers may react poorly to experimental packages having a version higher than 0.X. In these cases, experimental signals MAY version independently from stable signals, in order to retain a 0.X version number. When a signal becomes stable, the version MUST be bumped to match the other stable signals in the release.

### Major version bump

> Major version bumps MUST only occur when there is a breaking change to a stable interface, or the removal of a deprecated signal. Major version bumps MUST NOT occur for any other reason.

Reviewer 6: > Major version bumps MUST NOT occur for any other reason.

This is not required by SemVer by the way. SemVer requires "bump major version when ...." - not "you must not bump the major version unless ...". Do we need to do this?

A potential case for major version bump for "any other reason" - drop supporting end-of-life runtime. Some package manager supports checking the runtime requirement, but not all do that. For that reason, it's not uncommon to bump the major version only to drop the support for old runtime so that old runtime users wouldn't upgrade to the next version accidentally.

Author: Very good point. I will clarify that major version numbers SHOULD happen when we drop support for anything at all, and SHOULD NOT happen for other reasons. 

I received a number of questions along the lines of "if I add X does that bump the major version?" Which is why I wanted to emphasize that we bump the major version when we drop support, not when we add support.

    + [Contrib Stability](#contrib-stability)
    + [NOT DEFINED: Telemetry Stability](#not-defined-telemetry-stability)
  * [Deprecation](#deprecation)
  * [Removal](#removal)
  * [A note on replacing signals](#a-note-on-replacing-signals)
- [Version numbers](#version-numbers)
  * [Major version bump](#major-version-bump)
  * [Minor version bump](#minor-version-bump)
  * [Patch version bump](#patch-version-bump)
- [Long Term Support](#long-term-support)
  * [API support](#api-support)
  * [SDK Support](#sdk-support)
  * [Contrib Support](#contrib-support)
- [OpenTelemetry GA](#opentelemetry-ga)

<!-- tocstop -->

This document defines the stability guarantees offered by the OpenTelemetry clients, along with the rules and procedures for meeting those guarantees.

In this document, the terms "OpenTelemetry" and "language implementations" both specifically refer to the OpenTelemetry clients.
These terms do not refer to the specification or the Collector in this document.

Each language implementation MUST take these versioning and stability requirements, and produce a language-specific document which details how these requirements will be met.

> This document SHALL be placed in the root of each repo and named `VERSIONING`.

Reviewer 7: i suggest having a markdown file extension to be explicitly clear that the file is meant to be documentation and not a simple text file containing the latest version number
```suggestion
This document SHALL be placed in the root of each repo and named `VERSIONING.md`.
```

Author: I had that actually, but it was requested that markdown format not be required, so I removed it.

Provide maintainers a clear process for developing new, experimental signals alongside stable signals.
Different packages within the same release may have different levels of stability.
This means that an implementation wishing to release stable tracing today MUST ensure that experimental metrics are factored out in such a way that breaking changes to metrics API do not destabilize the trace API packages.

## Signal lifecycle

OpenTelemetry clients are structured around **signals**.
Each signal represents a coherent, stand-alone set of functionality.
Each signal follows a lifecycle.

![API Lifecycle](../internal/img/api-lifecycle.png)

### Lifecycle stages

**Experimental –** Breaking changes and performance issues MAY occur.
Components may not be feature-complete.
The experiment MAY be discarded and removed.

**Stable –** Stability guarantees, as defined by component type below, MUST now apply to all changes within this signal.
Long-term dependencies MAY now be taken against this signal.

**Deprecated –** this signal has been replaced, but MUST maintain the same stability guarantees.

**Removed -** a deprecated signal is no longer supported, and MUST be removed from the release.

> All signal components MAY become stable together, or MAY transition to stability component-by-component in the following order: API, Semantic Conventions, API Contrib, SDK, SDK Contrib.

Reviewer 4: What is the significance of this ordering? Does it mean, we cannot have stable SDK, *before* SemanticConventions are stable?

Author: Removing. The only actual rule is that the API must become stable before other components (or at the same time).

Package **version numbers** MAY include a suffix, such as -alpha, -beta, -rc, or -experimental, to differentiate stable and experimental packages.

### Stability

Once a signal component is marked as stable, the following rules MUST apply until the end of that signal’s existence.

#### API Stability

Backward-incompatible changes to API packages MUST NOT be made unless the major version number is incremented.
All existing API calls MUST continue to compile and function against all future minor versions of the same major version.

Languages which ship binary artifacts SHOULD offer [ABI compatibility](glossary.md#abi-compatibility) for API packages.

#### SDK Stability

Public portions of SDK packages MUST remain backwards compatible.
There are two categories of public features: **plugin interfaces** and **constructors**.
Examples of plugins include the SpanProcessor, Exporter, and Sampler interfaces.
Examples of constructors include configuration objects, environment variables, and SDK builders.

Languages which ship binary artifacts SHOULD offer [ABI compatibility](glossary.md#abi-compatibility) for SDK packages.

> #### Semantic Conventions Stability

Reviewer 8: Later in the document we say that Telemetry is not declared stable. However, at least of the Telemetry emitted depends on semantic conventions. So, in effect by saying Semantic Conventions are stable we are placing at least partial limitations on Telemetry as well.
I think we should also remove semantic conventions from initial stability guarantees.

Author: Ok I am convinced :)

Reviewer 8: @tedsuo do you plan to update this doc to remove Semantic Conventions Stability guarantees?

Author: Yes, sorry for the delay. I've updated the doc.

> Semantic Conventions MUST NOT be removed once they are stable.

Reviewer 9: Do we want to split the semconv into a separate package? 

Also, they are more related to telemetry and we are punting on telemetry stability in this PR?

Reviewer 9: I see this is alluded to below...it might be good to call it out along with the signals being packaged independently.

Author: Sure. I have some about this in the overview. Is there a another section which should be expanded to include these descriptions?

Reviewer 9: In the overview doc, it is explicitly mentioned in the API section that: 
`To manage different levels of stability, every signal has its own, independent API package.`

But not in the semantic conventions section. I think that's where I'm thinking it should be added.

Author: Based on discussion in other threads, I've removed references to this architecture proposal.

Older conventions MUST be marked as deprecated when they are replaced by newer conventions.

#### Contrib Stability

Plugins, instrumentation, and other contrib packages SHOULD be kept up to date and compatible with the latest versions of the API, SDK, and Semantic Conventions.
If a release of the API, SDK, or Semantic Conventions contains changes which are relevant to a contrib package, that package SHOULD be updated and released in a timely fashion.
The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Public portions of contrib packages (constructors, configuration, interfaces) SHOULD remain backwards compatible.

Languages which ship binary artifacts SHOULD offer [ABI compatibility](glossary.md#abi-compatibility) for contrib packages.

**Exception:** Contrib packages MAY break stability when a required downstream dependency breaks stability.
For example, a database integration may break stability if the required database client breaks stability.
However, it is strongly RECOMMENDED that older contrib packages remain stable.

> A new, incompatible version of an integration SHOULD be released as separate contrib package, rather than break the existing contrib package.

Reviewer 7: minor suggestion: add article
```suggestion
A new, incompatible version of an integration SHOULD be released as a separate contrib package, rather than break the existing contrib package.
```

> #### NOT DEFINED: Telemetry Stability

Reviewer 10: I don't have strong feelings about this, but was curious if we need to have `NOT DEFINED` here in the title, and as a result, in the TOC as well. This section _does_ define how we want to manage Telemetry stability - it's just that the plan is not complete yet. When there is a more-detailed plan, it will still belong in this section, and there won't be a change needed to the title of the section or to the TOC.

```suggestion
#### Telemetry Stability
```

Maybe this is an OTel documentation convention I'm not familiar with, in which case just ignore this suggestion.

**Telemetry stability guarantees are TBD.**

Changes to telemetry produced by OpenTelemetry instrumentation SHOULD avoid breaking analysis tools, such as dashboards and alerts.
However, it is not clear at this time what type of instrumentation changes (for example, adding additional spans and labels) would actually cause a breaking change.

**Until telemetry stability is defined, Contrib instrumentation MUST NOT be marked as stable.**

### Deprecation

Signals MAY eventually be replaced. When this happens, they are marked as deprecated.

Signals SHALL only be marked as deprecated when the replacement becomes stable.
Deprecated code MUST abide by the same support guarantees as stable code.

### Removal

Support is ended by the removal of a signal from the release.
The release MUST make a major version bump when this happens.

### A note on replacing signals

Note that we currently have no plans for creating a major version of OpenTelemetry past v1.0.

> For clarity, it is still possible to create new, backwards incompatible versions of existing signals without actually moving to v2.0 and breaking support.

Reviewer 10: ```suggestion
For clarity, it is still possible to create new, backwards-incompatible versions of existing signals without actually moving to v2.0 and breaking support.
```

For example, imagine we develop a new, better tracing API - let's call it AwesomeTrace.
We will never mutate the current tracing API into AwesomeTrace.
Instead, AwesomeTrace would be added as an entirely new signal which coexists and interoperates with the current tracing signal.
This would make adding AwesomeTrace a minor version bump, *not* v2.0.
v2.0 would mark the end of support for current tracing, not the addition of AwesomeTrace.
And we don't want to ever end that support, if we can help it.

This is not actually a theoretical example.
OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing.
We invented a new tracing API, but continue to support the old one.

## Version numbers

OpenTelemetry clients follow [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html), with the following clarifications.

OpenTelemetry clients have four components: API, SDK, Semantic Conventions, and Contrib.

> For the purposes of versioning, all code within a component MUST treated as if it were part of a single package, and versioned with the same version number,

Reviewer 10: ```suggestion
For the purposes of versioning, all code within a component MUST be treated as if it were part of a single package, and versioned with the same version number,
```

except for Contrib, which may be a collection of packages versioned separately.

> * API packages for all signals MUST version together, across all signals.

Reviewer 4: Did you mean to say "all stable signals" MUST version together?

Author: Ah, yes. I cleaned this language up.

Signals MUST NOT have separate version numbers.
There is one version number that applies to all signals that are included in the API release that is labeled with that particular version number.
* SDK packages for all signals MUST version together, across all signals.
Signals MUST NOT have separate version numbers.
There is one version number that applies to all signals that are included in the SDK release that is labeled with that particular version number.
* Semantic Conventions are a single package with a single version number.

> * Each contrib package MAY have it's own version number.

Reviewer 10: ```suggestion
* Each Contrib package MAY have its own version number.
```

> * The API, SDK, Semantic Conventions, and contrib components have independent version numbers.

Reviewer 10: ```suggestion
* The API, SDK, Semantic Conventions, and Contrib components have independent version numbers.
```

When a signal becomes stable, the version MUST be bumped to match the other stable signals in the release.

### Major version bump

Major version bumps MUST occur when there is a breaking change to a stable interface, the removal of a deprecated signal, or a drop in support for a language or runtime version.
Major version bumps SHOULD NOT occur for changes which do not result in a drop in support of some form.

### Minor version bump

Most changes to OpenTelemetry clients result in a minor version bump.

* New backward-compatible functionality added to any component.
* Breaking changes to internal SDK components.
* Breaking changes to experimental signals.
* New experimental signals are added.
* Experimental signals become stable.
* Stable signals are deprecated.

### Patch version bump

Patch versions make no changes which would require recompilation or potentially break application code.
The following are examples of patch fixes.

* Bug fixes which don't require minor version bump per rules above.
* Security fixes.
* Documentation.

Currently, the OpenTelemetry project does NOT have plans to backport bug and security fixes to prior minor versions of the SDK.

> Security and bug fixes MAY only applied to the latest minor version.

Reviewer 11: ```suggestion
Security and bug fixes MAY only be applied to the latest minor version.
```

We are committed to making it feasible for end users to stay up to date with the latest version of the OpenTelemetry SDK.

## Long Term Support

![long term support](../internal/img/long-term-support.png)

### API support

Major versions of the API MUST be supported for a minimum of **three years** after the release of the next major API version.
API support is defined as follows.

* API stability, as defined above, MUST be maintained.

> * A version of the SDK which supports the latest minor version of the last major version of the API will continue to be maintained during LTS.

Reviewer 10: ```suggestion
* A version of the SDK which supports the latest minor version of the last major version of the API MUST continue to be maintained during Long Term Support (LTS).
```

Bug and security fixes MUST be backported. Additional feature development is NOT RECOMMENDED.

* Contrib packages available when the API is versioned MUST continue to be maintained for the duration of LTS.
Bug and security fixes will be backported.
Additional feature development is NOT RECOMMENDED.

### SDK Support

SDK stability, as defined above, will be maintained for a minimum of **one year** after the release of the next major SDK version.

### Contrib Support

> Contrib stability, as defined above, will be maintained for a minimum of **one year** after after the release of the next major version of a contrib package.

Reviewer 7: minor suggestion: remove redundant word
```suggestion
Contrib stability, as defined above, will be maintained for a minimum of **one year** after the release of the next major version of a contrib package.
```

## OpenTelemetry GA

The term “OpenTelemetry GA” refers to the point at which OpenTracing and OpenCensus will be fully deprecated.

> The **minimum requirements** for declaring GA are as followed.

Reviewer 10: ```suggestion
The **minimum requirements** for declaring GA are as follows.
```
